src/group_theory/free_product.lean

```diff
@@ -0,0 +1,270 @@
+alphabet `Σ i, M i`. It's straightforward to define and easy to prove its universal property. But
+this answer is not completely satisfactory, because it's difficult to tell when two elements
+`x y : free_product M` are distinct since `free_product M` is defined as a quotient.
+
+The second, maximally efficient answer is given by `word M`. An element of `word M` is a word in the
+alphabet `Σ i, { m : M i // m ≠ 1 }`, where no adjacent letters can share an index `i`. Since we
+only work with reduced words, there is no need for quotienting, so it's easy to tell when two
+elements are distinct. However it's not obvious that this is even a monoid!
+
+We prove that every element of `free_product M` can be represented by a unique reduced word, i.e.
+`free_product M` and `word M` are equivalent types. This means that `word M` can be given a monoid
+structure, and it lets us tell when two elements of `free_product M` are distinct.
+
+There is also a completely tautological, maximally inefficient answer given by
+`algebra.category.Mon.colimits`. Whereas `free_product M` at least ensures that (any instance of)
+associativity holds by reflexivity, in this answer associativity holds because of quotienting. Yet
+another answer, which is constructively more satisfying, can be obtained by showing that
+`free_product.rel` is confluent.
+
+## References
+
+[van der Waerden, *Free products of groups*][MR25465]
+
+-/
+
+variables {ι : Type*} (M : Π i : ι, Type*) [Π i, monoid (M i)]
+
+/-- A relation on the free monoid on alphabet `Σ i, M i`, relating `⟨i, 1⟩` with `1` and
+`⟨i, x⟩ * ⟨i, y⟩` with `⟨i, x * y⟩`. -/
+inductive free_product.rel : free_monoid (Σ i, M i) → free_monoid (Σ i, M i) → Prop
+| of_one (i : ι) : free_product.rel (free_monoid.of ⟨i, 1⟩) 1
+| of_mul {i : ι} (x y : M i) : free_product.rel (free_monoid.of ⟨i, x⟩ * free_monoid.of ⟨i, y⟩)
+  (free_monoid.of ⟨i, x * y⟩)
+
+/-- The free product (categorical coproduct) of an indexed family of monoids. -/
+@[derive [monoid, inhabited]]
+def free_product : Type* := (con_gen (free_product.rel M)).quotient
+
+namespace free_product
+
+/-- The type of reduced words. A reduced word cannot contain a letter `1`, and no two adjacent
+letters can come from the same summand. -/
+def word : Type* := { l : list (Σ i, { m : M i // m ≠ 1 }) // (l.map sigma.fst).chain' (≠) }
+
+variable {M}
+
+/-- The inclusion of a summand into the free product. -/
+def of {i : ι} : M i →* free_product M :=
+{ to_fun   := λ x, con.mk' _ (free_monoid.of $ sigma.mk i x),
+  map_one' := (con.eq _).mpr (con_gen.rel.of _ _ (free_product.rel.of_one i)),
+  map_mul' := λ x y, eq.symm $ (con.eq _).mpr (con_gen.rel.of _ _ (free_product.rel.of_mul x y)) }
+
+lemma of_apply {i} (m : M i) : of m = con.mk' _ (free_monoid.of $ sigma.mk i m) := rfl
+
+variables {N : Type*} [monoid N]
+
+@[ext] lemma ext_hom (f g : free_product M →* N) (h : ∀ i, f.comp (of : M i →* _) = g.comp of) :
```

Comment:
There's a library note about partially applied ext lemmas you should reference here